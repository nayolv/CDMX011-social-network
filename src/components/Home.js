/* eslint-disable no-shadow */
/* eslint-disable eqeqeq */
import {
  db,
  getUser, sendEmail, signOut,
} from '../firebaseClient.js';
import { onNavigate } from '../main.js';

export const Home = () => {
  const user = getUser();
  const verification = user.emailVerified;
  const container = document.createElement('div');
  const html = `
  <header id=headerHome>
  <img id="logoHome" src="./sweatshirt.png">
  <h2>Trueque</h2>
  <a href ='' id='singOut' class="links">Cerrar Sesion</a>
  </header>
  <divPadre class='divPadre'>
  <h3> Te has logeado ${user ? user.email : ''} </h3>
  <p id=verificationMessage>${verification ? '' : 'Te enviamos un link a tu correo, verifica tu cuenta'}</p>
  <form id="wallForm">
  <div id="divPost">
  <input type="text" id="post" placeholder="¿Qué quieres publicar hoy?">
  <div id="divBtn">
  <a href="" id="btnPost" class="links">Publicar</a>
  </div>
  </div>
  </form>
  <div id="errorBackground">
  <p id= errorMessage></p>
  </div>
  <div id=postContainer></div>
  </divPadre>
  `;

  container.innerHTML = html;
  const postContainer = container.querySelector('#postContainer');
  const errorMessage = container.querySelector('#errorMessage');

  if (verification === false) {
    sendEmail()
      .then(() => {
        // MANDAMOS VERIFICACION AL CORREO DEL USUARIO
      });
  }
  // CERRAR SESION
  container.querySelector('#singOut').addEventListener('click', (e) => {
    e.preventDefault();
    signOut()
      .then(() => {
        onNavigate('/');
      })
      .catch((error) => {
        errorMessage.innerHTML = error.message;
        container.querySelector('#errorBackground').style.display = 'block';
      });
  });

  const docRef = db.collection('wallPost'); // CREANDO LA COLECCIÓN EN FB

  container.querySelector('#btnPost').addEventListener('click', (e) => {
    e.preventDefault();
    const publicaciones = document.querySelector('#post').value;
    const likes = [];
    const nameUser = user.email;
    if (publicaciones === '') {
      errorMessage.innerHTML = 'Por favor ingresa una publicación';
      container.querySelector('#errorBackground').style.display = 'block';
    } else {
      container.querySelector('#errorBackground').style.display = 'none';
      // AGREGA POST
      docRef.add({
        publicaciones,
        likes,
        nameUser,
      })
        .then(() => {
        // 'Document successfully written!
        })
        .catch((error) => {
          errorMessage.innerHTML = error.message;
          container.querySelector('#errorBackground').style.display = 'block';
        });
    }
    document.querySelector('#wallForm').reset();
  });

  docRef
    .onSnapshot((querySnapshot) => { // TRAEMOS LOS POST Y LOS AGREGAMOS EN TIEMPO REAL
      postContainer.innerHTML = '';
      querySnapshot.forEach((doc) => {
        const dataPost = doc.data();
        dataPost.id = doc.id;

        postContainer.innerHTML += `
        <section id="nameUserSection">
        <p id="nameUser">${dataPost.nameUser} escribió:</p>
        </section>
      <div id='contenedorPublicacion'>
      ${dataPost.publicaciones}
      <div id="btnsContenedor">
      <a href="" id='linkEdit' class="links" data-id='${dataPost.id}'>Editar</a>      
      <img id='btnDelete' src="./eliminar.png" data-id='${dataPost.id}'>
      <div class="likes">
      <img id="like" src="./corazon.png" data-id="${dataPost.id}">
      <span id="counter">${dataPost.likes.length}</span>
      </div>
      </div>
      </div>

      <div id="modalEdit" class="modalPadre">
      <div class="contenedorModal">
      <div class="closeDiv">
      <div class="confirmDiv">
      <p class="confirm">¿Editar esta publicación?</p>
      </div>
      <span class="close" id="closeEdit">&times;</span>
      </div>
      <input type="text" id="editInput">
      <button id="editBtnPost" data-id='${dataPost.id}'>Publicar</button>
      </div>
      </div>

      <div id="modalDelete" class="modalPadre">
      <div class="contenedorModal">
      <div class="closeDiv">
      <div class="confirmDiv">
      <p class="confirm">¿Borrar esta publicación?</p>
      </div>
      <span class="close" id="closeDelete">&times;</span>
      </div>
      <div id="deleteDiv"></div>
      <button id="deleteBtnPost" data-id='${dataPost.id}'>Borrar</button>
      </div>
      `;

        // BORRAR POST
        document.querySelectorAll('#btnDelete').forEach((btn) => { // SE RECORREN CON UN FOREACH
          btn.addEventListener('click', (e) => {
            document.getElementById('modalDelete').style.display = 'block';
            const target = e.target; // DELEGACION DE EVENTOS
            docRef.doc(target.dataset.id)
              .get()
              .then((doc) => {
                const dataDelete = doc.data();
                document.getElementById('deleteDiv').innerHTML = dataDelete.publicaciones;
              })
              .catch((error) => {
                errorMessage.innerHTML = error.message;
                container.querySelector('#errorBackground').style.display = 'block';
              });
            document.getElementById('deleteBtnPost').addEventListener('click', (e) => {
              const target = e.target;
              docRef.doc(target.dataset.id)
                .delete()
                .then(() => {
                // Document successfully deleted!
                }).catch((error) => {
                  errorMessage.innerHTML = error.message;
                  container.querySelector('#errorBackground').style.display = 'block';
                });
            });

            document.getElementById('closeDelete').addEventListener('click', () => {
              document.getElementById('modalDelete').style.display = 'none';
            });
          });
        });

        // EDITAR POST
        document.querySelectorAll('#linkEdit').forEach((btnE) => {
          btnE.addEventListener('click', (e) => {
            e.preventDefault();
            document.getElementById('modalEdit').style.display = 'block';

            const target = e.target;
            docRef.doc(target.dataset.id)
              .get()
              .then((doc) => {
                const dataEdit = doc.data();
                document.getElementById('editInput').value = dataEdit.publicaciones;
              })
              .catch((error) => {
                errorMessage.innerHTML = error.message;
                container.querySelector('#errorBackground').style.display = 'block';
              });

            document.getElementById('editBtnPost').addEventListener('click', () => {
              const target = e.target;
              const publicaciones = document.querySelector('#editInput').value;

              docRef.doc(target.dataset.id)
                .update({
                  publicaciones,
                })
                .then(() => {
                  // Document successfully updated!
                })
                .catch((error) => {
                  errorMessage.innerHTML = error.message;
                  container.querySelector('#errorBackground').style.display = 'block';
                });
            });
            document.getElementById('closeEdit').addEventListener('click', () => {
              document.getElementById('modalEdit').style.display = 'none';
            });
          });
        });

        // DAR LIKE POST
        document.querySelectorAll('#like').forEach((btn) => {
          btn.addEventListener('click', (e) => {
            const target = e.target;
            const docLikes = docRef.doc(target.dataset.id);
            docLikes.get()
              .then((doc) => {
                if (!doc.data().likes.includes(user.email)) {
                  docLikes
                    .update({
                      likes: firebase.firestore.FieldValue.arrayUnion(user.email),
                    });
                } else {
                  docLikes
                    .update({
                      likes: firebase.firestore.FieldValue.arrayRemove(user.email),
                    });
                }
              });
          });
        });
      });
    });

  return container;
};
