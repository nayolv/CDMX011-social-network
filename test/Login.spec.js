/**
 * @jest-environment jsdom
 */
import './globals/firebase';
import { Login } from '../src/components/Login.js';
import { render } from '../src/utils.js';
import { Register } from '../src/components/Register';
/* eslint-disable no-console */

describe('Testeando Login', () => {
  document.body.innerHTML = '<div id="root"></div>';

  test('debería renderizar', () => {
    const rootDiv = document.getElementById('root');
    const component = Login();
    render(rootDiv, component);

    expect(rootDiv.innerHTML).toMatchSnapshot();
  });

  test('Debe logear al usuario al hacer submit', () => {
    const mockLogin = jest.fn();
    mockLogin.mockImplementation(() => Promise.resolve());

    firebase.auth = jest.fn().mockImplementation(() => ({
      signInWithEmailAndPassword: mockLogin,
    }));
    const rootDiv = document.getElementById('root');
    const component = Login();
    render(rootDiv, component);

    const email = 'test@email.com';
    const password = '123456';

    document.getElementById('email').value = email;
    document.getElementById('password').value = password;

    document.getElementById('btnLogin').click();

    expect(mockLogin).toHaveBeenCalledWith(email, password);
  });
});
describe('Testeando Register', () => {
  document.body.innerHTML = '<div id="root"></div>';

  test('debería renderizar', () => {
    const rootDiv = document.getElementById('root');
    const component = Register();
    render(rootDiv, component);

    expect(rootDiv.innerHTML).toMatchSnapshot();
  });
  test('Debe crear al usuario al hacer submit', () => {
    const mockLogin = jest.fn();
    mockLogin.mockImplementation(() => Promise.resolve());

    firebase.auth = jest.fn().mockImplementation(() => ({
      createUserWithEmailAndPassword: mockLogin,
    }));
    const rootDiv = document.getElementById('root');
    const component = Register();
    render(rootDiv, component);

    const email = 'test@email.com';
    const password = '123456';

    document.getElementById('email').value = email;
    document.getElementById('password').value = password;

    document.getElementById('btnRegister').click();

    expect(mockLogin).toHaveBeenCalledWith(email, password);
  });
});

/* describe('Testeando Home', () => {
  document.body.innerHTML = '<div id="root"></div>';

  test('debería renderizar', () => {
    const rootDiv = document.getElementById('root');
    const component = Home();
    render(rootDiv, component);

    expect(rootDiv.innerHTML).toMatchSnapshot();
  });
}); */
