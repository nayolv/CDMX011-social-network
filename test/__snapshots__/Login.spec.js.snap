// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testeando Login debería renderizar 1`] = `
"<div> 
  <header>
  </header>
  <div class=\\"divPadre\\">
  <section id=\\"contenedorLogo\\">
  <img id=\\"logo\\" src=\\"./sweatshirt.png\\">
  <h1>Trueque</h1>
  <p id=\\"slogan\\">La comunidad más grande de intercambio de ropa.</p>
  </section>
  <div id=\\"errorBackground\\">
  <p id=\\"errorMessage\\"></p>
  </div>

  <main>
  <form class=\\"formulario\\">
  <section id=\\"sectionInputs\\">
  <input type=\\"email\\" id=\\"email\\" class=\\"inputs\\" placeholder=\\"example@email.com\\">
  <input type=\\"password\\" id=\\"password\\" class=\\"inputs\\" placeholder=\\"Contraseña\\">
  </section>
  <button id=\\"btnLogin\\">Iniciar sesión</button>
  <button id=\\"btnGoogle\\">Continua con Google</button>
  <a id=\\"textRegister\\">¿No tienes una cuenta?</a>
  <a href=\\"\\" id=\\"registerLink\\" class=\\"links\\">Registrate</a>
  </form>
  </main>
  </div>
  <footer>Trueque 2021</footer></div>"
`;

exports[`Testeando Register debería renderizar 1`] = `
"<div>
  <header>
  </header>
  <div class=\\"divPadre\\">
  <section id=\\"contenedorLogo\\">
  <img id=\\"logo\\" src=\\"./sweatshirt.png\\">
  <h1>Trueque</h1>
  <p id=\\"slogan\\">La comunidad más grande de intercambio de ropa.</p>
  </section>
  <div id=\\"errorBackground\\">
  <p id=\\"errorMessage\\"></p>
  </div>
  <main>
  <form class=\\"formulario\\">
  <section id=\\"sectionInputs\\">
  <input type=\\"email\\" id=\\"email\\" class=\\"inputs\\" placeholder=\\"example@email.com\\">
  <input type=\\"password\\" id=\\"password\\" class=\\"inputs\\" placeholder=\\"Contraseña\\">
  </section>
  <button id=\\"btnRegister\\">Registrate</button>
  <a id=\\"textRegister\\">¿Ya tienes una cuenta?</a>
  <a href=\\"\\" id=\\"loginLink\\" class=\\"links\\">Inicia sesión</a>
  <p id=\\"errorMessage\\"></p>
  </form>
  </main>
  </div>
  <footer>Trueque 2021</footer></div>"
`;
